new exception handler adoption

bkayes opened the pull request:
#### 079bd31e378144ae39bc940d07cfb8da5814f527
<pre>
new exception handler adoption
<a href="https://bugs.webkit.org/show_bug.cgi?id=269728">https://bugs.webkit.org/show_bug.cgi?id=269728</a>
<a href="https://rdar.apple.com/119256062">rdar://119256062</a>

Reviewed by Keith Miller.

This adopts the new hardened mach exception mechanism which was
introduced in darwin recently. See the radar for more details.

* Source/JavaScriptCore/llint/LLIntData.cpp:
* Source/JavaScriptCore/llint/LLIntData.h:
* Source/JavaScriptCore/runtime/MachineContext.h:
(JSC::MachineContext::setInstructionPointer):
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
(JSC::Wasm::trapHandler):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformMac.cmake:
* Source/WTF/wtf/PlatformRegisters.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::initMachExceptionHandlerThread):
(WTF::setExceptionPorts):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: <a href="https://commits.webkit.org/276090@main">https://commits.webkit.org/276090@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e18c7886f692d65fc714b002102f2e6b64bb7113

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43664 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22712 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46358 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39788 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45968 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26536 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20120 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36100 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44238 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19726 "Build is in progress. Recent messages:OS: Sonoma (14.3.1), Xcode: 15.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Running layout-tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37606 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17082 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17274 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38668 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1721 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/37123 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/simple-inline-stacktrace-with-catch.js.wasm-agressive-inline (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39820 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38975 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47858 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/43298 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18703 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15295 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42869 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20121 "Built successfully") | | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41545 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9724 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20302 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/50297 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19748 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10134 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1864,6 +1877,9 @@
 #if ENABLE(BLOCK_SET_EXCEPTION_PORTS)
             (with-filter (require-not (webcontent-process-launched))
                 (allow mach-message-send (kernel-mig-routine thread_set_exception_ports)))
```
You may have to remove the last closing parenthesis here.

pvollan replied:
Should `thread_set_exception_ports` only be allowed when `HAVE(MACH_HARDENED_EXCEPTIONS)` is not true?

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1839,10 +1839,23 @@
+        (with-filter (require-not (lockdown-mode))
+            (allow mach-message-send (kernel-mig-routine thread_set_exception_ports)))
+#end
+
```
This part may be unnecessary, since this should be covered by `ernel-mig-routines-blocked-in-lockdown-mode-avoid-telemetry`.

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1839,10 +1839,23 @@
+#if HAVE(MACH_HARDENED_EXCEPTIONS)
+    task_register_hardened_exception_handler
+    thread_adopt_exception_handler
+#endif
```
Should `thread_set_exception_ports` only appear in the `#else` of `HAVE(MACH_HARDENED_EXCEPTIONS)`?

kmiller68 commented on Source/JavaScriptCore/runtime/MachineContext.h:
```diff
@@ -457,6 +458,21 @@ inline void setInstructionPointer(PlatformRegisters& regs, CodePtr<CFunctionPtrT
 {
 #if USE(PLATFORM_REGISTERS_WITH_PROFILE)
     WTF_WRITE_PLATFORM_REGISTERS_PC_WITH_PROFILE(regs, value.taggedPtr());
+#elif USE(DARWIN_REGISTER_MACROS) && defined(EXCEPTION_STATE_IDENTITY_PROTECTED)
```
I think the way you've ordered this right now will only be enabled for watchOS. Since `USE_PLATFORM_REGISTERS_WITH_PROFILE` is enabled when `!defined(USE_PLATFORM_REGISTERS_WITH_PROFILE) && OS(DARWIN) && CPU(ARM64) && defined(__LP64__)` and AFAICT we don't override `USE_PLATFORM_REGISTERS_WITH_PROFILE` internally.

bkayes replied:
This is intentional - I redefined `WTF_WRITE_PLATFORM_REGISTERS_PC_WITH_PROFILE` in `PlatformRegisters.h` to account the "normal" case.

kmiller68 replied:
Ah, sorry missed that.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h:
```diff
@@ -37,4 +36,10 @@ inline void activateSignalingMemory() { }
 inline void prepareSignalingMemory() { }
 #endif // ENABLE(WEBASSEMBLY)
 
+class machExceptionSigningKey {
```
Classes are CamelCase in WebKit's style. Also, why have a class at all? I would just do all the registration in `prepareSignalingMemory()`

bkayes replied:
The problem is that you need the random number before calling `initMachExceptionHandlerThread`, and I couldn't figure out how to wrangle the build system to link the random number generator into a `WTF/*.cpp` file.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
```diff
@@ -38,15 +38,28 @@
 namespace {
 namespace WasmFaultSignalHandlerInternal {
-static constexpr bool verbose = false;
+static constexpr bool verbose = true;
```
Please revert.

kmiller68 commented on Source/WTF/wtf/PlatformMac.cmake:
```diff
@@ -117,7 +117,7 @@ add_custom_command(
     MAIN_DEPENDENCY mac/MachExceptions.defs
     WORKING_DIRECTORY ${WTF_DERIVED_SOURCES_DIR}
-    COMMAND mig -sheader MachExceptionsServer.h MachExceptions.defs
+    COMMAND mig -DMACH_EXC_SERVER_TASKIDTOKEN_STATE -sheader MachExceptionsServer.h MachExceptions2.defs
```
`MachExceptions2.defs` seems wrong.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -222,6 +245,106 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    const thread_state_t inState,
+    mach_msg_type_number_t inStateCount,
+    thread_state_t outState,
+    mach_msg_type_number_t* outStateCount);
```
I don't think you need this.

kmiller68 commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -129,9 +129,9 @@ WTF_EXPORT_PRIVATE void activateSignalHandlersFor(Signal);
-inline void initializeSignalHandling() { initMachExceptionHandlerThread(true); }
-inline void disableSignalHandling() { initMachExceptionHandlerThread(false); }
+WTF_EXPORT_PRIVATE void initMachExceptionHandlerThread(bool, uint32_t);
+inline void initializeSignalHandling(uint32_t signingKey = 0) { initMachExceptionHandlerThread(true, signingKey); }
```
This should just be a required argument and you'll need to add it to the non-Mach implementation below.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -110,8 +116,9 @@ inline void SignalHandlers::forEachHandler(Signal signal, const Func& func) cons
+uint32_t randomSigningKey = 0;
 
-void initMachExceptionHandlerThread(bool enable)
+void initMachExceptionHandlerThread(bool enable, __unused uint32_t signingKey)
```
WK style is to use `UNUSED_PARAM(signingKey)`.

kmiller68 commented on Source/JavaScriptCore/llint/LLIntData.cpp:
```diff
@@ -46,6 +46,7 @@ namespace LLInt {
 Opcode g_opcodeMap[numOpcodeIDs + numWasmOpcodeIDs] = { };
 Opcode g_opcodeMapWide16[numOpcodeIDs + numWasmOpcodeIDs] = { };
 Opcode g_opcodeMapWide32[numOpcodeIDs + numWasmOpcodeIDs] = { };
+CodePtr<NoPtrTag> presignedTrampoline = { };
```
This feels like it should be a new `OpaquePtrTag` or something. `NoPtrTag` implies it's an unsigned pointer.

kmiller68 replied:
Is this used?

kmiller68 reviewed (commented):
Some comments

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
```diff
@@ -38,12 +38,30 @@
+
+CodePtr<OpaquePtrTag> presignedTrampoline = { };
+
+#if defined(EXCEPTION_STATE_IDENTITY_PROTECTED)
```
I think these should probably be `#if defined(EXCEPTION_STATE_IDENTITY_PROTECTED) && EXCEPTION_STATE_IDENTITY_PROTECTED`

kmiller68 commented on Source/JavaScriptCore/llint/LLIntData.h:
```diff
@@ -30,6 +30,7 @@
 #include "JSCJSValue.h"
 #include "MacroAssemblerCodeRef.h"
 #include "Opcode.h"
+#include <wtf/PtrTag.h>
```
Seems like we don't need this here.

kmiller68 commented on Source/JavaScriptCore/llint/LLIntData.h:
```diff
@@ -47,6 +48,7 @@ extern "C" JS_EXPORT_PRIVATE JSC::Opcode g_opcodeMap[numOpcodeIDs + numWasmOpcod
 extern "C" JS_EXPORT_PRIVATE JSC::Opcode g_opcodeMapWide16[numOpcodeIDs + numWasmOpcodeIDs];
 extern "C" JS_EXPORT_PRIVATE JSC::Opcode g_opcodeMapWide32[numOpcodeIDs + numWasmOpcodeIDs];
 
+
```
Please revert.

kmiller68 commented on Source/WTF/wtf/PtrTag.h:
```diff
@@ -51,6 +51,7 @@ enum class PACKeyType {
 
 enum PtrTag : uintptr_t {
     NoPtrTag,
+    OpaquePtrTag = NoPtrTag,
```
I think we should put a comment here like:
// OpaquePtrTag is used for places where there is a tag but we intentionally have thrown away the diversifier. 
// For example, we register our WASM exception handler trampoline with a random key, pass the key to the OS, then throw the key away.
// So we have a tag but we can't verify it at runtime.

I think you also have to make this different from `NoPtrTag` because we assert in `CodePtr` that there's no tag on a `NoPtrTag` code.

bkayes replied:
I turned it into a `void *` because I think realistically this abstraction doesn't really fit super well. I think `void *` is sufficient for the purposes - do you agree?

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
```diff
@@ -38,12 +38,30 @@
+    // Sign the trampoline pointer using a random diversifier and stash it away before webcontent has started
+    presignedTrampoline = JSC::LLInt::getCodePtr<OpaquePtrTag>(wasm_throw_from_fault_handler_trampoline_reg_instance);
+    uint64_t diversifier = ptrauth_blend_discriminator((void *)(unsigned long)randomSigningKey, ptrauth_string_discriminator("pc"));
+    presignedTrampoline = CodePtr<OpaquePtrTag>::fromTaggedPtr(ptrauth_auth_and_resign(presignedTrampoline.taggedPtr(), ptrauth_key_process_independent_code, 0, ptrauth_key_process_independent_code, diversifier));
```
I think this is going to crash on debug builds because `CodePtr<OpaquePtrTag>::fromTaggedPtr` will do `assertIsTaggedWith<tag>(ptr);`. As `OpaquePtrTag == NoPtrTag`, `assertIsTaggedWith` will check there's no tag bits.

kmiller68 reviewed (commented):
Dropping these comments for now. I think this patch is going to crash on debug builds.

kmiller68 commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -124,18 +124,36 @@ struct SignalHandlers {
+inline exception_mask_t toMachMask(Signal signal)
+{
+    switch (signal) {
+    case Signal::AccessFault: return EXC_MASK_BAD_ACCESS;
+    case Signal::IllegalInstruction: return EXC_MASK_BAD_INSTRUCTION;
+    case Signal::FloatingPoint: return EXC_MASK_ARITHMETIC;
+    case Signal::Breakpoint: return EXC_MASK_BREAKPOINT;
```
Aren't these masks Mach only? So they won't exist on non-Mach Unix implementations.

kmiller68 commented on Source/JavaScriptCore/runtime/MachineContext.h:
```diff
@@ -457,12 +460,27 @@ inline void setInstructionPointer(PlatformRegisters& regs, CodePtr<CFunctionPtrT
 #endif
 }
+
+inline void setInstructionPointer(PlatformRegisters& regs, void *value)
```
WebKit style would be to have this be `void* value`

bkayes replied:
formatter caught the other ones, but missed this for some reason :)

kmiller68 commented on Source/JavaScriptCore/runtime/MachineContext.h:
```diff
@@ -44,6 +44,7 @@ template<typename T = void*> T framePointer(const PlatformRegisters&);
 inline CodePtr<PlatformRegistersLRPtrTag> linkRegister(const PlatformRegisters&);
 inline std::optional<CodePtr<PlatformRegistersPCPtrTag>> instructionPointer(const PlatformRegisters&);
 inline void setInstructionPointer(PlatformRegisters&, CodePtr<CFunctionPtrTag>);
+inline void setInstructionPointer(PlatformRegisters&, CodePtr<NoPtrTag>);
```
I think you need to update this declaration.

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -222,6 +231,93 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+kern_return_t catch_mach_exception_raise_state_identity_protected(
+    mach_port_t exceptionPort,
+    __unused uint64_t threadID,
+    __unused mach_port_t taskIDToken,
```
WebKit style is to use `UNUSED_PARAM`

kmiller68 commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -222,6 +231,93 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+        return KERN_SUCCESS;
+    }
+
+    return KERN_FAILURE;
```
This looks identical to `catch_mach_exception_raise_state`. Can we just push this all into one shared function?

pvollan reviewed (commented):
The sandbox changes look good.

kmiller68 commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -124,18 +124,36 @@ struct SignalHandlers {
 WTF_EXPORT_PRIVATE void addSignalHandler(Signal, SignalHandler&&);
 WTF_EXPORT_PRIVATE void activateSignalHandlersFor(Signal);
-
+WTF_EXPORT_PRIVATE void finalizeSignalHandlers(void);
```
You need to add a definition of this to `SignalsWin.cpp` for the windows build. Also `(void)` -> `()`

kmiller68 reviewed (approved):
r=me

Commit:
new exception handler adoption
https://bugs.webkit.org/show_bug.cgi?id=269728
rdar://119256062

Reviewed by Keith Miller.

This adopts the new hardened mach exception mechanism which was
introduced in darwin recently. See the radar for more details.

* Source/JavaScriptCore/llint/LLIntData.cpp:
* Source/JavaScriptCore/llint/LLIntData.h:
* Source/JavaScriptCore/runtime/MachineContext.h:
(JSC::MachineContext::setInstructionPointer):
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
(JSC::Wasm::trapHandler):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformMac.cmake:
* Source/WTF/wtf/PlatformRegisters.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::initMachExceptionHandlerThread):
(WTF::setExceptionPorts):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: https://commits.webkit.org/276090@main
